test/api/capi/capi_custom_type.cpp

```diff
@@ -0,0 +1,284 @@
+	const auto left_vector = duckdb_data_chunk_get_vector(input, 0);
+	const auto right_vector = duckdb_data_chunk_get_vector(input, 1);
+	const auto left_data = static_cast<float *>(duckdb_vector_get_data(duckdb_array_vector_get_child(left_vector)));
+	const auto right_data = static_cast<float *>(duckdb_vector_get_data(duckdb_array_vector_get_child(right_vector)));
+	const auto result_data = static_cast<float *>(duckdb_vector_get_data(duckdb_array_vector_get_child(output)));
+
+	for (idx_t i = 0; i < count; i++) {
+		for (idx_t j = 0; j < 3; j++) {
+			const auto idx = i * 3 + j;
+			result_data[idx] = left_data[idx] + right_data[idx];
+		}
+	}
+}
+
+static bool Vec3DToVarcharCastFunction(duckdb_function_info info, idx_t count, duckdb_vector input,
+                                       duckdb_vector output) {
+	const auto input_data = static_cast<float *>(duckdb_vector_get_data(duckdb_array_vector_get_child(input)));
+
+	for (idx_t i = 0; i < count; i++) {
+		const auto x = input_data[i * 3];
+		const auto y = input_data[i * 3 + 1];
+		const auto z = input_data[i * 3 + 2];
+		const auto res = StringUtil::Format("<%f, %f, %f>", x, y, z);
+		duckdb_vector_assign_string_element_len(output, i, res.c_str(), res.size());
+	}
+	return true;
+}
+
+static bool TryParseVec3D(char *str, idx_t len, float &x, float &y, float &z) {
+	char *end = str + len;
+	while (str < end && *str != '<') {
+		str++;
+	}
+	str++;
+	if (str >= end) {
+		return false;
+	}
+	x = std::strtof(str, &str);
+	if (str >= end || *str != ',') {
+		return false;
+	}
+	str++;
+	y = std::strtof(str, &str);
+	if (str >= end || *str != ',') {
+		return false;
+	}
+	str++;
+	z = std::strtof(str, &str);
+	if (str >= end || *str != '>') {
+		return false;
+	}
+	str++;
+	return str == end;
+}
+
+static bool Vec3DFromVarcharCastFunction(duckdb_function_info info, idx_t count, duckdb_vector input,
+                                         duckdb_vector output) {
+	const auto cast_mode = duckdb_cast_function_get_cast_mode(info);
+
+	// For testing purposes, check that we got the custom data
+	auto custom_data = reinterpret_cast<string *>(duckdb_cast_function_get_extra_info(info));
+	REQUIRE(*custom_data == "foobar");
+
+	const auto input_data = static_cast<duckdb_string_t *>(duckdb_vector_get_data(input));
+	auto output_data = static_cast<float *>(duckdb_vector_get_data(duckdb_array_vector_get_child(output)));
+
+	bool success = true;
+	for (idx_t i = 0; i < count; i++) {
+		auto x = 0.0f, y = 0.0f, z = 0.0f;
+		auto str = input_data[i];
+		auto str_len = duckdb_string_t_length(str);
+		char *str_ptr = duckdb_string_is_inlined(str) ? str.value.inlined.inlined : str.value.pointer.ptr;
+
+		// yes, sscanf is not safe, but this is just a test
+		if (TryParseVec3D(str_ptr, str_len, x, y, z)) {
+			// Success
+			output_data[i * 3] = x;
+			output_data[i * 3 + 1] = y;
+			output_data[i * 3 + 2] = z;
+		} else {
+			// Error
+			if (cast_mode == DUCKDB_CAST_TRY) {
```

Comment:
Maybe we can have a helper function for this similar to how we handle this in the C++ functions, e.g.:

```cpp
void (*duckdb_cast_function_set_row_error)(duckdb_function_info info, const char *error, idx_t index, duckdb_vector output);
```

Reply:
So this would basically invalidate the validity at the current row and set the error message (if not already set) in the same call? I worry that if the user is doing some sort of string formatting to create the error message they are going to perform that work unnecessarily for each invalid row after the first one in the case of a try cast, but maybe that's something they can guard against themselves (by checking the cast mode and passing nullptr for subsequent errors) if that's the case.

Reply:
Yes, in the C++ layer templating takes care of that - but this would require some user handling if they want to avoid that cost. That being said I still think it's cleaner than having to essentially duplicate this code in every cast function - and also less error prone.

Reply:
Alright, do I keep the original `cast_function_set_error()`?

Reply:
I think that's fine yeah, there might be situations where we want to return errors from a `try_cast` as well.